### PR TITLE
Add ability to create/validate invalid foreign keys in Postgres

### PR DESCRIPTION
**Pull request by travisofthenorth**
### Summary

In Postgres, adding foreign keys can cause significant downtime because the transaction needs to acquire some very heavy locks on the table being altered as well as the table being referenced. To illustrate, if I wanted to add a foreign key on my `addresses` table referencing my `users` table:
```
myapp=# BEGIN;
BEGIN
myapp=# ALTER TABLE addresses ADD CONSTRAINT "fk_rails_48c9e0c5a2" FOREIGN KEY ("user_id") REFERENCES "users" ("id");
ALTER TABLE
myapp=# SELECT locktype, relation::regclass, mode, transactionid AS tid, virtualtransaction AS vtid, pid, granted FROM pg_locks;
   locktype    |          relation          |        mode         |   tid   |  vtid   |  pid  | granted
---------------+----------------------------+---------------------+---------+---------+-------+---------
 relation      | pg_locks                   | AccessShareLock     |         | 2/80438 | 56984 | t
 relation      | users_pkey                 | AccessShareLock     |         | 2/80438 | 56984 | t
 relation      | index_addresses_on_user_id | AccessShareLock     |         | 2/80438 | 56984 | t
 relation      | addresses_pkey             | AccessShareLock     |         | 2/80438 | 56984 | t
 virtualxid    |                            | ExclusiveLock       |         | 2/80438 | 56984 | t
 relation      | addresses                  | AccessShareLock     |         | 2/80438 | 56984 | t
 relation      | addresses                  | AccessExclusiveLock |         | 2/80438 | 56984 | t
 transactionid |                            | ExclusiveLock       | 3919702 | 2/80438 | 56984 | t
 relation      | users                      | AccessShareLock     |         | 2/80438 | 56984 | t
 relation      | users                      | RowShareLock        |         | 2/80438 | 56984 | t
 relation      | users                      | AccessExclusiveLock |         | 2/80438 | 56984 | t
(11 rows)
```

...my transaction acquires an `AccessExclusiveLock` on `users` which is extremely detrimental on a high-traffic table, esp. when Postgres performs a potentially lengthy query to validate the check.

On the other hand, I can take a two-step approach which significantly reduces this burden; by introducing an invalid constraint in one transaction and validating it in another, the locks acquired are much less restrictive:
```
myapp=# BEGIN;
BEGIN
myapp=# ALTER TABLE addresses ADD CONSTRAINT "fk_rails_48c9e0c5a2" FOREIGN KEY ("user_id") REFERENCES "users" ("id") NOT VALID;
ALTER TABLE
myapp=# COMMIT;
COMMIT
myapp=# BEGIN;
BEGIN
myapp=# ALTER TABLE addresses VALIDATE CONSTRAINT "fk_rails_48c9e0c5a2";
ALTER TABLE
myapp=# SELECT locktype, relation::regclass, mode, transactionid AS tid, virtualtransaction AS vtid, pid, granted FROM pg_locks;
   locktype    |          relation          |           mode           |   tid   |  vtid   |  pid  | granted
---------------+----------------------------+--------------------------+---------+---------+-------+---------
 relation      | users_pkey                 | AccessShareLock          |         | 2/80443 | 56984 | t
 relation      | index_addresses_on_user_id | AccessShareLock          |         | 2/80443 | 56984 | t
 relation      | addresses_pkey             | AccessShareLock          |         | 2/80443 | 56984 | t
 relation      | addresses                  | AccessShareLock          |         | 2/80443 | 56984 | t
 relation      | users                      | AccessShareLock          |         | 2/80443 | 56984 | t
 relation      | users                      | RowShareLock             |         | 2/80443 | 56984 | t
 relation      | pg_locks                   | AccessShareLock          |         | 2/80443 | 56984 | t
 virtualxid    |                            | ExclusiveLock            |         | 2/80443 | 56984 | t
 relation      | addresses                  | ShareUpdateExclusiveLock |         | 2/80443 | 56984 | t
 transactionid |                            | ExclusiveLock            | 3919706 | 2/80443 | 56984 | t
(10 rows)
```

The first transaction acquires the same `AccessExclusiveLock` on the `users` table, but "the potentially-lengthy initial check to verify that all rows in the table satisfy the constraint is skipped" (source: [postgres docs](https://www.postgresql.org/docs/9.3/static/sql-altertable.html)). Subsequently, the validation step does not block reads or writes on the `users` table. 💯 

So, this PR introduces two things:

- The ability to create invalid foreign keys by specifying the option `valid: false`
- A `validate_foreign_key` method (which takes the same variety of params as the other foreign key methods) to validate a foreign key

I've heard rumors about this being on the roadmap for the Postgres team, i.e. skipping the check if the table being altered is empty and marking the constraint valid. In any case, perhaps someday this will be more easily achieved with built-in Postgres, but for now it's an issue.

### CODE REVIEW

**Comment by travisofthenorth** [comment]
Please advise on any places to add test coverage. I was having a bit of difficulty identifying where to add tests for certain things.

**Comment by travisofthenorth** [comment]
Looks like `create_table` does allow foreign key creation in the same statement, so I am adding support for that.

**Comment by travisofthenorth** [comment]
@maclover7 is there a process I should be following for getting some 👀 on this PR? I believe I have some spec failures due to the `bundler` update issue, but nothing related to this code.

**Comment by maclover7** [comment]
@travisofthenorth I added the `needs feedback` label, so hopefully someone will review shortly 😬

**Comment by travisofthenorth** [comment]
@schneems @kaspth just pinging random members at this point...sorry about that. Just wondering if I can get any feedback on this?

**Comment by kaspth** [comment]
Sorry don't have the Postgres experience to vet this.

@kamipo you've been doing wonders with Active Record, are you interested in giving this PR a review? 😊

**Comment by kamipo** [comment]
`VALIDATE CONSTRAINT` works not only foreign key but also CHECK constraint.
And also Oracle have similar feature as `NOVALIDATE` (novalidate constraint to existing records).

So I prefer the naming here:

* `supports_invalid_foreign_keys?` -> `supports_validate_constraints?`
  * what supports `VALIDATE CONSTRAINT` means that allowing `NOVALIDATE` constraint.
* `validate_foreign_key` -> `validate_constraint`
* `not_valid?` -> `novalidate?`

**Comment by travisofthenorth** [comment]
@kamipo updated. A couple things to mention:

- I kept the `validate_foreign_key` method. It's nice because it takes the same signature as the existing methods, and a developer doesn't need to specify the constraint name directly.
- I added a `validate_constraint` statement which takes the table and constraint names.
- I noticed you made some conflicting updates upstream. Please advise it this is still ok: https://github.com/travisofthenorth/rails/blob/b458c7733404da4727569ff684fb73ffd379ab35/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L6-L11

**Comment by travisofthenorth** [comment]
Ping @kamipo. Any other feedback?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: If an adapter doesn't support novalidate constraint, these `valid?` and `novalidate?` returns incorrect result.
How about it?

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
index 4332318a8e..5127d6bbe9 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
@@ -58,12 +58,8 @@ def custom_primary_key?
         options[:primary_key] != default_primary_key
       end
 
-      def valid?
-        options[:valid] == true
-      end
-
-      def novalidate?
-        options[:valid] == false
+      def validate?
+        options.fetch(:validate, true)
       end
 
       def defined_for?(to_table_ord = nil, to_table: nil, **options)
```

**Comment by travisofthenorth** [comment]
@kamipo feedback addressed. I left `valid?` in there because it's helpful for querying existing constraints.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: I prefer `validate?` predicate method.

```ruby
      def validate?
        options.fetch(:validate, true)
      end
```

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: We can remove `valid?` to use `options[:validate]`.

Comment (travisofthenorth) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: That doesn't make much sense to me. At this point, the foreign key has already been created. I'd expect a `valid?` or `validated?` method to describe the status of an existing foreign key and `options[:validate]` to be used only during creation.

Either way, I'll make the change in a subsequent commit.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
Comment (travisofthenorth) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: This is a little strange to me because you could pass in `validate: nil` or `validate: 'foo'` and hence you wouldn't be getting back a boolean from this method.

**Comment by travisofthenorth** [comment]
@kamipo feedback addressed. I'm a bit hesitant about the usage of `validate?` for existing FKs but I will defer to you.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: There are other predicate methods that do not return boolean (e.g. `connection.in_use?`, `connection.strict_mode?`.
I think it should be named according to how it is used.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: Agreed at your point. How about `alias validated? validate?`?

Comment (travisofthenorth) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: Done

**Comment by travisofthenorth** [comment]
Added `validated?` as an alias

**Comment by kamipo** [comment]
Looks good to me 👍

**Comment by travisofthenorth** [comment]
Squashed and rebased. Thanks for the all the feedback @kamipo!

**Comment by travisofthenorth** [comment]
@kamipo @maclover7 is there any chance this PR could make it into 5.1? I saw that beta1 was released and it made me curious.

**Comment by travisofthenorth** [comment]
@kamipo just wanted to check in. Do you need anything else from me here?

**Comment by travisofthenorth** [comment]
@kamipo now that Rails 5.1 is out there, is it easier to get this in? This feature is becoming increasingly important for one of the apps I contribute to.

**Comment by kamipo** [comment]
I can't merge PRs, so you'll have to wait for someone else from the Committers or Core teams to review this PR.

**Comment by travisofthenorth** [comment]
K, thanks for letting me know. Sorry for the noise!

**Comment by travisofthenorth** [comment]
Rebased on master and resolved a conflict.

**Comment by matthewd** [comment]
I'm a bit worried about how much stuff we're adding to the abstract layer, for a feature that only Postgres supports. Is any of that avoidable? :confused:

**Comment by schneems** [comment]
I would like to see Rails promote more best practices in regards to database use. Foreign key and other constraints are a huge part of this. On one hand it is catering to a specific database. On the other hand, should users of that DB get a worse experience just because ALL databases don't support that feature?

I know it's not covered in this PR but validation race conditions are a huge issue. Recently I had a validation cause over 80% of ALL load on my postgres database, and I had no idea. Here's the PR explaining the issue and fixing it https://github.com/codetriage/codetriage/pull/573.

It would like to see Rails continue to play better with postgres in the future.

@matthewd  RE: size. Half of the PR is tests. I agree that we want to keep the abstract layer to not get too bloated. Any alternative implementation ideas?

**Comment by travisofthenorth** [comment]
@matthewd it seemed unavoidable but it was my first time digging into the adapter code so I could be wrong. I'm happy to make changes so let me know if you have any ideas.

**Comment by travisofthenorth** [comment]
@matthewd I took another look at it. The latest commit moves most of the implementation into postgres-specific classes. I left some code in the `ForeignKeyDefinition` struct in abstract because it felt weird re-implementing that. Let me know how it looks now.

**Comment by travisofthenorth** [comment]
Rebased on master and squashed to 1 commit. @matthewd please advise if you want to see any more changes in this PR, otherwise I think it should be good to go.

**Comment by travisofthenorth** [comment]
@kamipo @matthewd just rebased on master. Any chance this could make it into 5.2?

**Review by kamipo — APPROVED** [review]

**Comment by matthewd** [comment]
Sorry I didn't come back to this sooner, especially after you did a great job of addressing my concern about the split between abstract vs postgres adapters. 👍🏻

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb
Comment (thizzle) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: This comment doesn't match the method it's documenting.

Comment (bogdanvlviv) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: It was fixed by 70c96b4423abacecb1116184c8ea4d4662b809f8

**Pull request merged**
